core/src/main/java/org/apache/calcite/rel/metadata/RelMdPredicates.java

```diff
@@ -666,6 +666,29 @@ static class JoinConditionBasedPredicateInference {
       equivalence = BitSets.closure(equivalence);
     }
 
+    /**
+     * As RexPermuteInputsShuttle, with one exception. When visiting an inputRef,
+     * it will replace the type of the InputRef with the type found in the input fields,
+     * instead of keeping the original type. This is used within
+     * when generating the Left/RightInferredPredicates, to avoid nullability mismatches
+     * between the types of the join and the types of the inputs.
+     */
+    private class TypeChangingRexPermuteInputsShuttle
+        extends org.apache.calcite.rex.RexPermuteInputsShuttle {
```

Comment:
Is there a situation where the behavior of `TypeChangingRexPermuteInputsShuttle.visitInputRef` is *not* correct for the regular `RexPermuteInputsShuttle`? 
If not, then maybe this should just be the behavior of `RexPermuteInputsShuttle`

Reply:
I tried doing something like this in `RexPermuteInputsShuttle` 

```
if (fields.isEmpty()) {
      return new RexInputRef(
          target,
          local.getType());
    } else {
      return new RexInputRef(
          target,
          fields.get(target).getType());
    }
```
(Note that the if statement is needed because it's valid to create a `RexPermuteInputsShuttle` without any backing fields.)

This caused several tests to fail with index out of bound errors. The reason wasn't obvious to me. I'll do a bit of poking around and see if I can figure out why.

Reply:
Maybe there are situations where fields is not empty, but also not the same size as the number of input columns?
I could see something like this working:
```java
RelDataType dataType = target < fields.size() ? local.getType() : fields.get(target).getType();
```